### PR TITLE
[release-4.10] Bug 2097735: Fix loadBalancerServiceAnnotationsChanged check and update

### PR DESCRIPTION
**Pull request by Miciah**
This is a manual cherry-pick of #783.  #707 introduced a conflict that required resolution.  

#### `loadBalancerServiceAnnotationsChanged`: Fix update

Fix `loadBalancerServiceAnnotationsChanged` to use the provided `annotations` argument when updating annotations.  Before this change, `loadBalancerServiceAnnotationsChanged` used the `annotations` argument when checking whether annotations had changed, but then it used `managedLoadBalancerServiceAnnotations` when setting the annotations in the updated service object.

When checking whether the service is upgradeable for the purpose of setting the ingresscontroller's "Upgradeable" status condition, `loadBalancerServiceTagsModified` calls `loadBalancerServiceAnnotationsChanged` with an `annotations` argument that includes the "service.beta.kubernetes.io/aws-load-balancer-additional-resource-tags" annotation key, which is not in `managedLoadBalancerServiceAnnotations`.  As a result, `loadBalancerServiceAnnotationsChanged` would report that the service was not upgradeable, but then the status condition message would not show that the "aws-load-balancer-additional-resource-tags" annotation had changed.

* `pkg/operator/controller/ingress/load_balancer_service.go` (`loadBalancerServiceAnnotationsChanged`): Use the provided `annotations` argument instead of using `managedLoadBalancerServiceAnnotations` when updating annotations.
* `pkg/operator/controller/ingress/load_balancer_service_test.go` (`TestLoadBalancerServiceAnnotationsChanged`): New test.  Verify that `loadBalancerServiceAnnotationsChanged` behaves correctly when an unmanaged annotation is updated or when a managed annotation is added, updated, or deleted.

#### `loadBalancerServiceAnnotationsChanged`: Fix check

Fix `loadBalancerServiceAnnotationsChanged` to treat nil and empty annotations maps as equal.  Before this commit, the comparison erroneously flagged an empty map and a nil map as unequal.  As a result, the ingress operator could erroneously report that an ingresscontroller with a service with no annotations was not upgradeable.

* `pkg/operator/controller/ingress/load_balancer_service.go` (`loadBalancerServiceAnnotationsChanged`): Use a `for` loop instead of go-cmp to check whether any managed annotations have changed.
* `pkg/operator/controller/ingress/load_balancer_service_test.go` (`TestLoadBalancerServiceAnnotationsChanged`): Add test cases for empty/nil current/expected annotations.

### CODE REVIEW

**Comment by Miciah** [comment]
/retitle  [release-4.10] Bug 2097735: Fix loadBalancerServiceAnnotationsChanged check and update

**Comment by Miciah** [comment]
```
 === RUN   TestHAProxyTimeouts
    haproxy_timeouts_test.go:153: Expected value for "timeout server" to be "90s", got "90s\n120s"
    haproxy_timeouts_test.go:178: deleted ingresscontroller haproxy-timeout
--- FAIL: TestHAProxyTimeouts (37.93s)
=== RUN   TestHAProxyTimeoutsRejection
    haproxy_timeouts_test.go:290: Expected value for "timeout server" to be "30s", got "30s\n120s"
    haproxy_timeouts_test.go:318: deleted ingresscontroller haproxy-timeout-rejection
--- FAIL: TestHAProxyTimeoutsRejection (37.89s) 
```
These failures are being tracked by [BZ#2100630](https://bugzilla.redhat.com/show_bug.cgi?id=2100630).

**Comment by Prashanth684** [comment]
is this good to be merged? this is affecting the 4.10 CI jobs

**Comment by Miciah** [comment]
> is this good to be merged? this is affecting the 4.10 CI jobs

We can't merge it until [BZ#2100630](https://bugzilla.redhat.com/show_bug.cgi?id=2100630) is fixed.

**Comment by Miciah** [comment]
I've posted https://github.com/openshift/cluster-ingress-operator/pull/792.

**Comment by Miciah** [comment]
`TestConfigurableRouteNoConsumingUserNoRBAC` failed, must-gather failed, and a worker node was unschedulable.  
/test e2e-aws-operator

**Comment by candita** [comment]
This PR is mentioned as blocking 4.10 nightly for 3 days, so I tried to investigate it.  Documented in:
https://coreos.slack.com/archives/CFDM5CQMN/p1656698903916469

TLDR: a must-gather 10m timeout was added on May 6: https://github.com/openshift/release/commit/6173489a76aaccf6842fe023365b3e632cc2c53a
The last time the test worked for us, May 5, the must-gather took 40 minutes:
https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_cluster-ingress-operator/755/pull-ci-openshift-cluster-ingress-operator-release-4.10-e2e-aws-operator/1522214355762941952/build-log.txt

**Comment by Miciah** [comment]
This has been blocking nightlies for 3 weeks.  Examining the last 5 e2e-aws-operator failures:

* [1542971493829316608](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-ingress-operator/784/pull-ci-openshift-cluster-ingress-operator-release-4.10-e2e-aws-operator/1542971493829316608) failed because of unschedulable nodes and the must-gather timeout.
* [1542923639324676096](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-ingress-operator/784/pull-ci-openshift-cluster-ingress-operator-release-4.10-e2e-aws-operator/1542923639324676096) failed because of `UpdatingPrometheusK8SFailed`, unschedulable nodes, and the must-gather timeout.
* [1542511915886972928](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-ingress-operator/784/pull-ci-openshift-cluster-ingress-operator-release-4.10-e2e-aws-operator/1542511915886972928) failed because of `UpdatingPrometheusK8SFailed` and the must-gather timeout.
* [1542274566984830976](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-ingress-operator/784/pull-ci-openshift-cluster-ingress-operator-release-4.10-e2e-aws-operator/1542274566984830976) failed because of `TestConfigurableRouteNoConsumingUserNoRBAC`, unschedulable nodes, and the must-gather timeout.
* [1542159989899005952](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-ingress-operator/784/pull-ci-openshift-cluster-ingress-operator-release-4.10-e2e-aws-operator/1542159989899005952) failed because of the must-gather timeout.

We were told we cannot increase the must-gather timeout, and as that is the only consistent cause of failure, I'm going to override e2e-aws-operator in order to unblock 4.10 nightlies.

/override e2e-aws-operator

**Comment by Miciah** [comment]
Got `PrometheusOperatorWatchErrors`.
/test e2e-aws

**Comment by Miciah** [comment]
The ingress operator failed to publish the wildcard DNS record:
```
2022-07-01T23:27:30.612Z	ERROR	operator.dns_controller	dns/controller.go:192	failed to publish DNS record to zone	{"record": {"dnsName":"*.apps.ci-op-bn8c37rl-a9b88.origin-ci-int-gce.dev.rhcloud.com.","targets":["35.238.58.143"],"recordType":"A","recordTTL":30}, "dnszone": {"id":"ci-op-bn8c37rl-a9b88-ps7xh-private-zone"}, "error": "Post \"https://dns.googleapis.com/dns/v1/projects/openshift-gce-devel-ci/managedZones/ci-op-bn8c37rl-a9b88-ps7xh-private-zone/changes?alt=json&prettyPrint=false\": oauth2: cannot fetch token: 400 Bad Request\nResponse: {\"error\":\"invalid_grant\",\"error_description\":\"Invalid grant: account not found\"}"}
```
/test e2e-gcp-serial

**Comment by Miciah** [comment]
`PrometheusOperatorWatchErrors` failed.
/test e2e-aws
